File: app/android/core/src/main/kotlin/com/neuropulse/core/session/SessionHistoryStore.kt
package com.neuropulse.core.session

import com.neuropulse.core.common.KeyValueStore
import java.time.LocalDate
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

// Local, display-only record of completed sessions. Port of iOS
// SessionHistoryStore.swift.
//
// PRIVACY (UHDR boundary): this store NEVER persists raw EEG/HRV waveform
// data — only display-aggregated metrics already shown post-session.
//
// Exact completion timestamps are UHDR-class data (CLAUDE.md §5.1 — "session
// timestamps → UHDR") and are NOT stored here. Only the calendar day is
// persisted (sessionDay, "yyyy-MM-dd" local timezone) plus a monotonic
// insertionIndex for same-day ordering (NP-PRIV-ANALYSIS-002 MEDIUM-07).
//
// edfSessionId is in-memory only (@Transient) — the hub session epoch is a
// Unix-ms timestamp and must not reach persisted storage.

@Serializable
data class SessionRecord(
    val id: String,
    val protocolName: String,
    // Calendar day the session completed — "yyyy-MM-dd", device local timezone.
    val sessionDay: String,
    // Monotonic insertion counter; tiebreaks same-day records newest-first.
    val insertionIndex: Int,
    val durationSeconds: Double,
    val averageCoherenceScore: Float? = null, // 0.0–10.0; null when protocol had no HRV
    val rmssdMilliseconds: Int? = null,       // integer only — never a raw RR series
    val impedancePassCount: Int,              // 0–8 EEG electrodes that passed
    // Deliberately excluded from serialization — UHDR-class (see header comment).
    @Transient val edfSessionId: Long? = null,
)

// Post-session summary handed in by the session flow (in-memory only).
data class CompletedSessionSummary(
    val protocolName: String,
    val completedAt: LocalDate,
    val durationSeconds: Int,
    val averageCoherenceScore: Float?,
    val rmssdMilliseconds: Int?,
    val impedancePassCount: Int,
    val edfSessionId: Long? = null,
    // Closed-loop adaptation events (UHDR-class). Populated by the live post-session path;
    // NEVER persisted (see store header), so a summary reconstructed from a history record has
    // an empty list — the Adaptive Adjustments card then shows its empty state, as on iOS.
    val adaptationEvents: List<AdaptationEvent> = emptyList(),
) {
    companion object {
        /** Reconstruct a display summary from a persisted history row (no adaptation events). */
        fun fromRecord(record: SessionRecord): CompletedSessionSummary = CompletedSessionSummary(
            protocolName = record.protocolName,
            completedAt = runCatching { LocalDate.parse(record.sessionDay) }.getOrDefault(LocalDate.MIN),
            durationSeconds = record.durationSeconds.toInt(),
            averageCoherenceScore = record.averageCoherenceScore,
            rmssdMilliseconds = record.rmssdMilliseconds,
            impedancePassCount = record.impedancePassCount,
            edfSessionId = record.edfSessionId,
            adaptationEvents = emptyList(),
        )
    }
}

class SessionHistoryStore(
    private val store: KeyValueStore,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    companion object {
        const val STORAGE_KEY = "np.session.history"
        const val MAX_RECORDS = 100

        // Companion-scoped so it exists before the instance init block runs
        // (a body-declared val would still be null during init { load() }).
        private val newestFirst = Comparator<SessionRecord> { a, b ->
            if (a.sessionDay != b.sessionDay) {
                b.sessionDay.compareTo(a.sessionDay)
            } else {
                b.insertionIndex.compareTo(a.insertionIndex)
            }
        }
    }

    var records: List<SessionRecord> = emptyList()
        private set

    private var nextInsertionIndex = 0

    init {
        load()
    }

    /** Append a completed session, persist, trim, and keep newest-first. */
    fun record(session: CompletedSessionSummary) {
        val new = SessionRecord(
            id = java.util.UUID.randomUUID().toString(),
            protocolName = session.protocolName,
            sessionDay = session.completedAt.toString(), // ISO yyyy-MM-dd
            insertionIndex = nextInsertionIndex,
            durationSeconds = session.durationSeconds.toDouble(),
            averageCoherenceScore = session.averageCoherenceScore,
            rmssdMilliseconds = session.rmssdMilliseconds,
            impedancePassCount = session.impedancePassCount,
            edfSessionId = session.edfSessionId,
        )
        nextInsertionIndex += 1
        records = (records + new).sortedWith(newestFirst).take(MAX_RECORDS)
        persist()
    }

    fun clearHistory() {
        records = emptyList()
        nextInsertionIndex = 0
        persist()
    }

    // ── Persistence ──────────────────────────────────────────────────────

    private fun load() {
        val blob = store.getString(STORAGE_KEY) ?: return
        // Corrupt blob is non-fatal — start clean rather than crash on launch.
        records = runCatching {
            json.decodeFromString(ListSerializer(SessionRecord.serializer()), blob)
                .sortedWith(newestFirst)
        }.getOrDefault(emptyList())
        nextInsertionIndex = (records.maxOfOrNull { it.insertionIndex } ?: -1) + 1
    }

    private fun persist() {
        store.putString(
            STORAGE_KEY,
            json.encodeToString(ListSerializer(SessionRecord.serializer()), records),
        )
    }
}

File: app/ios/NeurOne/Session/SessionHistoryStore.swift
import Foundation
import OSLog

// Local, display-only record of completed sessions (ISC-48, ISC-49).
//
// PRIVACY (UHDR boundary): this store NEVER persists raw EEG/HRV waveform data.
// It holds only display-aggregated metrics already shown post-session:
// average coherence score, RMSSD as an integer, and the EEG impedance pass count.
// These are the same values surfaced live in SessionView. Raw biological time
// series remains exclusively in the on-device encrypted UHDR partition and is
// never copied into UserDefaults.
//
// completedAt is UHDR-class data (CLAUDE.md §5.1 — "session timestamps → UHDR")
// and is NOT stored here. Only the calendar day is persisted (sessionDay, "yyyy-MM-dd"
// in local timezone). Day granularity is sufficient for all display purposes and
// is safe to store in UserDefaults, which is included in iCloud/iTunes backups and
// MDM-accessible. Exact timestamps go to the encrypted UHDR partition only.
// (NP-PRIV-ANALYSIS-002 MEDIUM-07)
//
// The EDF+ file itself (full waveform archive) lives in Documents/UHDR and is
// referenced here only by its hub-assigned session ID (edfSessionID) — the
// store never decodes or stores its contents.

// One persisted history row. Codable for UserDefaults JSON storage.
struct SessionRecord: Identifiable, Codable, Equatable {
    var id: UUID = UUID()
    var protocolName: String
    // Calendar day the session completed — "yyyy-MM-dd" in device local timezone.
    // Day granularity only; exact timestamps are UHDR-class data and not stored here.
    var sessionDay: String
    // Monotonic insertion counter. Tiebreaks same-day records in newest-first order
    // after timestamp coarsening removes sub-day precision.
    var insertionIndex: Int
    var durationSeconds: Double
    var averageCoherenceScore: Float?   // 0.0–10.0; nil when protocol had no HRV
    var rmssdMilliseconds: UInt16?      // integer only — never a raw RR series
    var impedancePassCount: Int         // 0–8 EEG electrodes that passed
    // edfSessionID is held in-memory only — intentionally excluded from Codable
    // serialization. gatt.session.epoch is a Unix-ms timestamp (UHDR-class,
    // CLAUDE.md §5.1) and must not be persisted to the backed-up UserDefaults store.
    // The download button is available in the live post-session window; after app
    // restart the hub presents historical sessions via Mode 4 on reconnect.
    var edfSessionID: UInt32? = nil

    private enum CodingKeys: String, CodingKey {
        case id, protocolName, sessionDay, insertionIndex, durationSeconds
        case averageCoherenceScore, rmssdMilliseconds, impedancePassCount
        // edfSessionID deliberately omitted — see comment above
    }

    // Shared "yyyy-MM-dd" formatter in device local timezone.
    // Used on store write and display.
    static let dayFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()
}

@MainActor
final class SessionHistoryStore: ObservableObject {

    // Newest-first ordering is maintained on every mutation.
    @Published private(set) var records: [SessionRecord] = []

    private let defaults: UserDefaults
    private let storageKey = "np.session.history"
    private static let maxRecords = 100
    // Monotonically increasing; tiebreaks same-day records newest-first.
    private var nextInsertionIndex = 0

    private static let logger = Logger(subsystem: "com.neurone.app", category: "SessionHistoryStore")

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        load()
    }

    // Append a completed session, persist, trim, and keep newest-first.
    func record(_ session: CompletedSessionSummary) {
        let new = SessionRecord(
            protocolName: session.protocolName,
            sessionDay: SessionRecord.dayFormatter.string(from: session.completedAt),
            insertionIndex: nextInsertionIndex,
            durationSeconds: Double(session.durationSeconds),
            averageCoherenceScore: session.averageCoherenceScore,
            rmssdMilliseconds: session.rmssdMilliseconds,
            impedancePassCount: session.impedancePassCount,
            edfSessionID: session.edfSessionID
        )
        nextInsertionIndex += 1
        records.insert(new, at: 0)
        records.sort(by: Self.newestFirst)
        if records.count > Self.maxRecords {
            records = Array(records.prefix(Self.maxRecords))
        }
        persist()
    }

    func clearHistory() {
        records = []
        nextInsertionIndex = 0
        persist()
    }

    // MARK: - Persistence

    private func load() {
        guard let data = defaults.data(forKey: storageKey) else { return }
        do {
            var decoded = try JSONDecoder().decode([SessionRecord].self, from: data)
            decoded.sort(by: Self.newestFirst)
            records = decoded
            nextInsertionIndex = (records.map(\.insertionIndex).max() ?? -1) + 1
        } catch {
            // Corrupt blob is non-fatal — start clean rather than crash on launch.
            Self.logger.error("Failed to decode session history: \(String(describing: error))")
            records = []
        }
    }

    private func persist() {
        do {
            let data = try JSONEncoder().encode(records)
            defaults.set(data, forKey: storageKey)
        } catch {
            Self.logger.error("Failed to encode session history: \(String(describing: error))")
        }
    }

    private static func newestFirst(_ a: SessionRecord, _ b: SessionRecord) -> Bool {
        if a.sessionDay != b.sessionDay { return a.sessionDay > b.sessionDay }
        return a.insertionIndex > b.insertionIndex
    }
}
